pkg/connector/kairosdb.go

```diff
@@ -0,0 +1,568 @@
+		string(data))
+
+	if err = json.Unmarshal(data, &JSONquery); err != nil {
+		return fmt.Errorf("kairosdb[%s]: unable to unmarshal JSON data: %s", connector.name, err)
+	}
+
+	for _, q := range JSONquery["queries"] {
+		for _, r := range q["results"] {
+			metricName := r.Name
+			aggregator := matchAggregatorPattern(connector.aggregators, metricName)
+			if aggregator == nil {
+				aggregator = connector.defAggregator
+			}
+
+			for _, t := range connector.srcTags {
+				if _, ok := r.Tags[t]; !ok {
+					continue
+				}
+				var sc uint64
+				for _, sourceName := range r.Tags[t] {
+					if _, ok := connector.series[sourceName]; !ok {
+						connector.series[sourceName] = make(map[string]kairosdbSeriesEntry)
+					}
+					connector.series[sourceName][metricName] = kairosdbSeriesEntry{
+						tag: t,
+						source: sourceName,
+						metric: metricName,
+						aggregator: aggregator,
+					}
+					outputChan <- &catalog.Record{
+						Origin:    originName,
+						Source:    sourceName,
+						Metric:    metricName,
+						Connector: connector,
+					}
+					sc++
+				}
+				logger.Log(logger.LevelInfo, "connector", "kairosdb[%s]: %d sources for `%s'", connector.name, sc, metricName)
+				if aggregator != nil {
+					a, _ := json.Marshal(aggregator);
+					logger.Log(logger.LevelInfo, "connector", "kairosdb[%s]: `%s' applied to `%s'",
+						connector.name, string(a), metricName)
+				}
+				break
+			}
+		}
+	}
+	return nil
+}
+
+func kairosdbBuildJSONQuery(query *plot.Query, kairosdbSeries map[string]map[string]kairosdbSeriesEntry) ([]byte, error) {
+	type PlotsQuery struct {
+		StartAbsolute int64               `json:"start_absolute"`
+		EndAbsolute   int64               `json:"end_absolute"`
+		Metrics       []MetricsQueryEntry `json:"metrics"`
+	}
+
+	q := PlotsQuery{StartAbsolute: query.StartTime.Unix() * 1000,
+			EndAbsolute: query.EndTime.Unix() * 1000}
+
+	for _, series := range query.Series {
+		entry := kairosdbSeries[series.Source][series.Metric]
+		m := MetricsQueryEntry{
+			Name: entry.metric,
+			Tags: map[string][]string{entry.tag: []string{entry.source}},
+		}
+		// catch `json:"aggregators,omitempty"` to avoid "aggregators": [null]
+		if entry.aggregator != nil {
+			m.Aggregators = []interface{}{entry.aggregator}
+		} else {
+			m.Aggregators = nil
+		}
+		q.Metrics = append(q.Metrics, m)
+	}
+	jsonData, err := json.Marshal(q)
+	if err != nil {
+		return nil, err
+	}
+
+	return jsonData, nil
+}
+
+func kairosdbExtractPlots(query *plot.Query, kairosdbSeries map[string]map[string]kairosdbSeriesEntry, kairosdbPlots []MetricsQueryResponse) ([]plot.Series, error) {
+	var resultSeries []plot.Series
+
+	for _, kairosdbPlot := range kairosdbPlots {
+
+		// REVIEW: is there a better approach to retrieve target?
+		var target string = ""
+		for _, series := range query.Series {
+
+			entry := kairosdbSeries[series.Source][series.Metric]
+
+			// REVIEW (KairosDB API): more than one result possible?
```

Comment:
Well, from what I've observed querying KairosDB even when requesting multiple series at once the API returns an array `queries` of objects featuring an array `results` yielding only one element array containing the datapoints for the series requested – so it looks like processing only the first element of each `results` array is fine.

Test query (2 series) : http://pastebin.com/8RLAnkKr
Response : http://pastebin.com/KHHzvgWe